### PR TITLE
KAFKA-3590: Handle not-enough-replicas errors when writing to offsets topic

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala [hunk 1]
```diff
@@ -202,36 +202,45 @@ class GroupMetadataManager(val brokerId: Int,
...
+
+            error(s"Appending metadata message for group ${group.groupId} generation $generationId failed due to " +
+              s"${statusError.exceptionName}, returning UNKNOWN error code to the client")
+
+            Errors.UNKNOWN
 
-          status.errorCode
+          case other =>
+            error(s"Appending metadata message for group ${group.groupId} generation $generationId failed " +
+              s"due to unexpected error code: ${statusError.code}")
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala]: Why are we not using `exceptionName` here?

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala [hunk 2]
```diff
@@ -101,14 +101,50 @@ class GroupMetadataManagerTest {
...
+    var maybeError: Option[Errors] = None
+    def callback(error: Errors) {
+      maybeError = Some(error)
     }
 
     val delayedStore = groupMetadataManager.prepareStoreGroup(group, Map.empty, callback)
     groupMetadataManager.store(delayedStore)
-    assertEquals(Errors.NONE.code, errorCode.get)
+    assertTrue(maybeError.isDefined)
+    assertEquals(Errors.NONE, maybeError.get)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala]: You will get better errors if you replace both lines above with:

``` scala
assertEquals(Some(Errors.NONE), maybeError)
```

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala [hunk 3]
```diff
@@ -101,14 +101,50 @@ class GroupMetadataManagerTest {
...
+  @Test
+  def testStoreGroupErrorMapping() {
+    assertStoreGroupErrorMapping(Errors.NONE, Errors.NONE)
+    assertStoreGroupErrorMapping(Errors.UNKNOWN_TOPIC_OR_PARTITION, Errors.GROUP_COORDINATOR_NOT_AVAILABLE)
+    assertStoreGroupErrorMapping(Errors.NOT_ENOUGH_REPLICAS, Errors.GROUP_COORDINATOR_NOT_AVAILABLE)
+    assertStoreGroupErrorMapping(Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND, Errors.GROUP_COORDINATOR_NOT_AVAILABLE)
+    assertStoreGroupErrorMapping(Errors.NOT_LEADER_FOR_PARTITION, Errors.NOT_COORDINATOR_FOR_GROUP)
+    assertStoreGroupErrorMapping(Errors.MESSAGE_TOO_LARGE, Errors.UNKNOWN)
+    assertStoreGroupErrorMapping(Errors.RECORD_LIST_TOO_LARGE, Errors.UNKNOWN)
+    assertStoreGroupErrorMapping(Errors.INVALID_FETCH_SIZE, Errors.UNKNOWN)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala:123]: How come we use `UNKNOWN` for `MESSAGE_TOO_LARGE` for example?

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala [hunk 4]
```diff
@@ -101,14 +101,50 @@ class GroupMetadataManagerTest {
...
+
+    var maybeError: Option[Errors] = None
+    def callback(error: Errors) {
+      maybeError = Some(error)
+    }
+
+    val delayedStore = groupMetadataManager.prepareStoreGroup(group, Map.empty, callback)
+    groupMetadataManager.store(delayedStore)
+    assertTrue(maybeError.isDefined)
+    assertEquals(expectedError, maybeError.get)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala]: Same as elsewhere, if you assert that the result is a `Some(...)`, you get a better error than the `assertTrue(foo.isDefined)`. And save a line too. ;)

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala [hunk 5]
```diff
@@ -130,14 +166,15 @@ class GroupMetadataManagerTest {
...
+    var maybeError: Option[Errors] = None
+    def callback(error: Errors) {
+      maybeError = Some(error)
     }
 
     val delayedStore = groupMetadataManager.prepareStoreGroup(group, Map(memberId -> Array[Byte]()), callback)
     groupMetadataManager.store(delayedStore)
-    assertEquals(Errors.NONE.code, errorCode.get)
+    assertTrue(maybeError.isDefined)
+    assertEquals(Errors.NONE, maybeError.get)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala]: Similar as the other messages above.

**Review by ijuma — CHANGES_REQUESTED** [review]
Minor suggestions and a couple of questions, looks good otherwise.

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala]: Ack

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala:123]: I was a little surprised to see this also. @guozhangwang can probably clarify, but this is basically an unrecoverable error for the consumer, so I think the idea is to return an error to the client which will cause it to crash, but which doesn't expose the internal storage mechanism.

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala]: Can't think of a good reason. I'll change it.

**Comment by hachikuji** [comment]
@ijuma Ready for another look when you have a chance.

**Comment by ijuma** [comment]
LGTM

File: core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala (hunk 3 above)
Comment (guozhangwang) [reply core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala:123]: @hachikuji Shouldn't this be mapped to `Errors.INVALID_COMMIT_OFFSET_SIZE`?

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala:123]: @guozhangwang This case is for persistence of group metadata.

Comment (guozhangwang) [reply core/src/test/scala/unit/kafka/coordinator/GroupMetadataManagerTest.scala:123]: Got it. LGTM.

**Comment by hachikuji** [comment]
Merged to trunk and 0.10.1.

**Pull request closed without merging**